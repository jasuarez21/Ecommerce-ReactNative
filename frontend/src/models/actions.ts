import { Moto } from "./moto";
import { User } from "./user";

export type MotoActions = 
    | {type: 'GET_MOTOS'; data: Array<Moto>}
    | {type: 'ADD_TO_CART'; moto: Moto}
    | {type: 'LOAD_CART';}
    | {type: 'DELETE_CART'; moto: Moto}
    | {type: 'LOG_IN'; data: User}
    | {type: 'LOG_OUT'; data: User}
    | {type: 'SIGN_UP'; data: User}
    | {type: 'UPDATE_CART'}
    | {type: 'DELETE_ACCOUNT'}
    | {type: 'ERROR'}