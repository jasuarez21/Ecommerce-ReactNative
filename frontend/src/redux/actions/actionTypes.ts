const actionTypes = {
  GET_MOTOS: 'GET_MOTOS',
  LOG_IN: 'LOG_IN',
  LOG_OUT: 'LOG_OUT',
  SIGN_UP: 'SIGN_UP',
  ADD_TO_CART: 'ADD_TO_CART',
  LOAD_CART: 'LOAD_CART',
  DELETE_CART: 'DELETE_CART',
  UPDATE_CART: 'UPDATE_CART',
  SAVE_PRODUCT: 'SAVE_PRODUCT',
  DELETE_ACCOUNT: 'DELETE_ACCOUNT',
  ERROR: 'ERROR'
};

export default actionTypes;
