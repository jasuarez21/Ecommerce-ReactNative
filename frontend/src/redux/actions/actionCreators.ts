import { User } from './../../models/user';
import axios from 'axios';
import {URL_MOTOS, URL_LOGIN, URL_LOGOUT, URL_SIGNUP, URL_DELETE_ACCOUNT, URL_SAVE_PRODUCTS} from '@env';
import actionTypes from './actionTypes';
import {Moto} from '../../models/moto';
import {Dispatch} from 'redux';
import {MotoActions} from '../../models/actions';
import { Order } from '../../models/order';

export function GetMotos() {
  return async (dispatch: Dispatch) => {
      try {   
      const {data} = await axios.get(URL_MOTOS);
      dispatch({
        type: actionTypes.GET_MOTOS,
        data,
      });
    } catch (error) {
      dispatch({
        type: actionTypes.ERROR,
      });
    }
  }
}



export function login(email: string, password: string) {
  return async (dispatch: Dispatch) => {
      try {   
        const {data} = await axios.post(URL_LOGIN, {email, password});
        dispatch({
          type: actionTypes.LOG_IN,
          data,
        });
    } catch (error) {
      dispatch({
        type: actionTypes.ERROR,
      });
    }
  }
}

export function logout() {
  return async (dispatch: Dispatch) => {
      try {   
        const {data} = await axios.post(URL_LOGOUT);
        dispatch({
          type: actionTypes.LOG_OUT,
          data,
        });
    } catch (error) {
      dispatch({
        type: actionTypes.ERROR,
      });
    }
  }
}


export function signUp(email: string, password: string, alias: string, avatar: string) {
  return async (dispatch: Dispatch) => {
      try {   
        const {data} = await axios.post(URL_SIGNUP, {email, password, alias, avatar});
        dispatch({
          type: actionTypes.SIGN_UP,
          data,
        });
    } catch (error) {
      dispatch({
        type: actionTypes.ERROR,
      });
    }
  }
}

export function addToCart(moto: Moto): MotoActions {
  return {
    type: 'ADD_TO_CART',
    moto
  };
}

export function loadCart(): MotoActions {
  return {
    type: 'LOAD_CART'
  };
}

export function deleteFromCart(moto: Moto): MotoActions {
  return {
    type: 'DELETE_CART',
    moto,
  };
}

export function reloadCart(moto: Array<Moto>){
  return async (dispatch: Dispatch) => {
    try {   
      for(let i = 0; i < moto.length; i++){
        await axios.post(`${URL_MOTOS}/${moto[i]._id}`, moto[i]);
        dispatch({
          type: actionTypes.UPDATE_CART
        });
      }
  } catch (error) {
    dispatch({
      type: actionTypes.ERROR,
    });
  }
}
}

export function saveOrder(user: User, cart: Order, token: string){
  user.cart.push(cart);
  const config = { 
    headers: {
        authorization: `Bearer ${token}`,
    }
  }
    return async (dispatch: Dispatch) => {
      try {   
          await axios.post(URL_SAVE_PRODUCTS, {_id: user._id, cart: user.cart}, config);
          dispatch({
            type: actionTypes.SAVE_PRODUCT
          });
    } catch (error) {
      dispatch({
        type: actionTypes.ERROR,
      });
    }
  }
}

export function deleteAccount(user: User, token: string){
    return async (dispatch: Dispatch) => {
      try {   
          await axios.delete(URL_DELETE_ACCOUNT, {headers: {
              authorization: `Bearer ${token}`,
          },
              data:{_id: user._id}
          });
          dispatch({
            type: actionTypes.DELETE_ACCOUNT
          });
    } catch (error) {
      dispatch({
        type: actionTypes.ERROR,
      });
    }
  }
}