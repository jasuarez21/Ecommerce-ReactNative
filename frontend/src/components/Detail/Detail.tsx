import React from 'react';
import {useSelector, useDispatch} from 'react-redux';
import {RootState} from '../../redux/store/index'
import {Moto} from '../../models/moto';
import {
  Text,
  Image,
  TouchableHighlight,
  ImageBackground,
  View,
  StyleSheet,
} from 'react-native';
import { addToCart } from '../../redux/actions/actionCreators';

const Detail = ({route}: RootState) => {
  const motos = useSelector((store:any) => store.motos);
  const dispatch = useDispatch();
  const {motoSelected} = route.params;
  const motoDetail = motos.filter((moto: Moto) => motoSelected === moto._id);
  return (
    <>
    <ImageBackground
      source={{uri: motoDetail[0].picture}}
      style={styles.background}>
        <View style={styles.containerLogo}>
           <Text style={styles.detailMoto}>{motoDetail[0].marca} {motoDetail[0].modelo}</Text>
        </View>
        <Text style={styles.price}>{motoDetail[0].precio}â‚¬</Text>
      <View style={styles.information}>
        <Text style={styles.textinfo}>
            {motoDetail[0].descripcion}
        </Text>
        <Text style={styles.textinfo}>Marca: {motoDetail[0].marca}</Text>
        <Text style={styles.textinfo}>Modelo: {motoDetail[0].modelo}</Text>
        <Text style={styles.textinfo}>CC: {motoDetail[0].cc}</Text>
        <Text style={styles.textinfo}>Stock: {
        motoDetail[0].stock >= 1 ? (
            motoDetail[0].stock
        ): (
            <Text>No hay stock</Text>
        )
        }</Text>
      </View>
      { motoDetail[0].stock !== 0 ? (
        <TouchableHighlight
          style={styles.buttonCart}
          onPress={() => (
            motoDetail[0].stock -1,
            dispatch(addToCart(motoDetail[0])))}>
           <Image
                testID= 'addButton'
                style={styles.cart}
                source={{
                  uri: 'https://i.postimg.cc/GpkSCktY/icons8-buy-52.png',
                }}
              />
        </TouchableHighlight>
      ) : (
        <View style={styles.buttonCartDisabled}>
            <Image
                style={styles.cart}
                source={{
                  uri: 'https://i.postimg.cc/jq7ZrsJr/icons8-out-of-stock-30.png',
                }}
              />
        </View>
      )
      
      }
    </ImageBackground>
        </>
  );
};
const styles = StyleSheet.create({
  background: {
    width: 450,
    height: 500,
  },
  noStock: {
    color: 'white'
  },
  buttonCart: {
    width: 110,
    height: 30,
    position: 'absolute',
    borderRadius: 10,
    textAlign: 'center',
    marginLeft: 150
  },
  buttonCartDisabled: {
    width: 110,
    height: 30,
    position: 'absolute',
    borderRadius: 10,
    textAlign: 'center',
    marginLeft: 150
  },
  containerLogo: { 
    backgroundColor: 'black',
    height: 65,
  },
  cart: {
    width: 35,
    height: 35,
    marginTop: 15,
    marginLeft: 50
  },
  information: {
    backgroundColor: 'black',
    opacity: 0.8,
    padding: 10,
    marginTop: 280,
    height: 500,
    width: 405,
    borderTopLeftRadius: 25,
    borderTopRightRadius: 25,
  },
  detailMoto: {
    fontSize: 24,
    color: '#FFB800',
    marginTop: 20,
    marginLeft: 20,
  },
  price:{
    fontSize: 24,
    color: '#FFB800',
    marginTop: 20,
    marginLeft: 330,
    position: 'absolute'
  },
  textinfo: {
    margin: 10,
    fontSize: 18,
    color: 'white',
  },
  logo: {
    width: 300,
    height: 42,
    marginTop: 15,
    marginBottom: 50
  }
});

export default (Detail);