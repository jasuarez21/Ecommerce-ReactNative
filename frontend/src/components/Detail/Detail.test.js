import { render, fireEvent } from '@testing-library/react-native';
import React from 'react';
import ConfigureStore from 'redux-mock-store';
import * as actions from '../../redux/actions/actionCreators';
import { Provider } from 'react-redux';
import Detail from './Detail';

describe('When invoked a Detail component', () => {
    const mockStore = ConfigureStore();
    const store = mockStore({
        motos: [{
              _id: '2',
              marca: 'Aprilia',
              modelo: 'SR',
              cc: 50, 
              precio: 550,
              descripcion: 'Moto comoda para la carretera',
              picture: 'img.com'
          }
        ]
    });
    let route;
    let motos;
    beforeEach(() => {
        jest.spyOn(actions, 'addToCart').mockReturnValueOnce({ type: '' });
        route = {
            params: {
                motoSelected : '2'
            }
        }
      });
    test('Should render a moto with detail', () => {
        const detail = render(<Provider store={store}><Detail route={route} motos={motos} /></Provider>)
        expect(detail).toMatchSnapshot();
    })
})

describe('When invoked a Detail component', () => {
    const mockStore = ConfigureStore();
    const store = mockStore({
        motos: [{
              _id: '2',
              marca: 'Aprilia',
              modelo: 'SR',
              cc: 50, 
              precio: 550,
              descripcion: 'Moto comoda para la carretera',
              picture: 'img.com',
              stock: 3
          }
        ]
    });
    let route;
    let motos;
    beforeEach(() => {
        jest.spyOn(actions, 'addToCart').mockReturnValueOnce({ type: '' });
        route = {
            params: {
                motoSelected : '2'
            }
        }
      });
    test('Should render a moto with detail', () => {
        const detail = render(<Provider store={store}><Detail route={route} motos={motos} /></Provider>)
        expect(detail).toMatchSnapshot();
    })
    test('Then navigation.navigate addbuton is invoked', () => {
        const { getByTestId } = render(<Provider store={store}><Detail route={route} motos={motos} /></Provider>);
        const image = getByTestId('addButton');
        fireEvent.press(image);
        expect(actions.addToCart).toHaveBeenCalled();
      });
})