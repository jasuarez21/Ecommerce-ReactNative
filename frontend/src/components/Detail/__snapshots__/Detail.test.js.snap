// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When invoked a Detail component Should render a moto with detail 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "height": 500,
      "width": 450,
    }
  }
>
  <Image
    source={
      Object {
        "uri": "img.com",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": 500,
          "width": 450,
        },
        undefined,
      ]
    }
  />
  <View
    style={
      Object {
        "backgroundColor": "black",
        "height": 65,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFB800",
          "fontSize": 24,
          "marginLeft": 20,
          "marginTop": 20,
        }
      }
    >
      Aprilia
       
      SR
    </Text>
  </View>
  <Text
    style={
      Object {
        "color": "#FFB800",
        "fontSize": 24,
        "marginLeft": 330,
        "marginTop": 20,
        "position": "absolute",
      }
    }
  >
    550
    €
  </Text>
  <View
    style={
      Object {
        "backgroundColor": "black",
        "borderTopLeftRadius": 25,
        "borderTopRightRadius": 25,
        "height": 500,
        "marginTop": 280,
        "opacity": 0.8,
        "padding": 10,
        "width": 405,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 18,
          "margin": 10,
        }
      }
    >
      Moto comoda para la carretera
    </Text>
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 18,
          "margin": 10,
        }
      }
    >
      Marca: 
      Aprilia
    </Text>
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 18,
          "margin": 10,
        }
      }
    >
      Modelo: 
      SR
    </Text>
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 18,
          "margin": 10,
        }
      }
    >
      CC: 
      50
    </Text>
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 18,
          "margin": 10,
        }
      }
    >
      Stock: 
      <Text>
        No hay stock
      </Text>
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "borderRadius": 10,
        "height": 30,
        "marginLeft": 150,
        "position": "absolute",
        "textAlign": "center",
        "width": 110,
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "https://i.postimg.cc/GpkSCktY/icons8-buy-52.png",
        }
      }
      style={
        Object {
          "height": 35,
          "marginLeft": 50,
          "marginTop": 15,
          "width": 35,
        }
      }
      testID="addButton"
    />
  </View>
</View>
`;

exports[`When invoked a Detail component Should render a moto with detail 2`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "height": 500,
      "width": 450,
    }
  }
>
  <Image
    source={
      Object {
        "uri": "img.com",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": 500,
          "width": 450,
        },
        undefined,
      ]
    }
  />
  <View
    style={
      Object {
        "backgroundColor": "black",
        "height": 65,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFB800",
          "fontSize": 24,
          "marginLeft": 20,
          "marginTop": 20,
        }
      }
    >
      Aprilia
       
      SR
    </Text>
  </View>
  <Text
    style={
      Object {
        "color": "#FFB800",
        "fontSize": 24,
        "marginLeft": 330,
        "marginTop": 20,
        "position": "absolute",
      }
    }
  >
    550
    €
  </Text>
  <View
    style={
      Object {
        "backgroundColor": "black",
        "borderTopLeftRadius": 25,
        "borderTopRightRadius": 25,
        "height": 500,
        "marginTop": 280,
        "opacity": 0.8,
        "padding": 10,
        "width": 405,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 18,
          "margin": 10,
        }
      }
    >
      Moto comoda para la carretera
    </Text>
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 18,
          "margin": 10,
        }
      }
    >
      Marca: 
      Aprilia
    </Text>
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 18,
          "margin": 10,
        }
      }
    >
      Modelo: 
      SR
    </Text>
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 18,
          "margin": 10,
        }
      }
    >
      CC: 
      50
    </Text>
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 18,
          "margin": 10,
        }
      }
    >
      Stock: 
      3
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "borderRadius": 10,
        "height": 30,
        "marginLeft": 150,
        "position": "absolute",
        "textAlign": "center",
        "width": 110,
      }
    }
  >
    <Image
      source={
        Object {
          "uri": "https://i.postimg.cc/GpkSCktY/icons8-buy-52.png",
        }
      }
      style={
        Object {
          "height": 35,
          "marginLeft": 50,
          "marginTop": 15,
          "width": 35,
        }
      }
      testID="addButton"
    />
  </View>
</View>
`;
